### PR TITLE
ExecutionStack change

### PR DESCRIPTION
**Pull request by RoiEXLab**
## Overview
This renames the fields of ExecutionStack to make checkstyle happy and refactors the class a bit.

## Functional Changes
None.

## Manual Testing Performed
I verified I could play a game normally, the tests were added for the refactred version of the class, but I verified they failed for my "incorrect implementation" (see commit history), and passed for the old implementation.

### CODE REVIEW

File: game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java [hunk 1]
```diff
@@ -28,37 +29,36 @@
...
   void execute(final IDelegateBridge bridge) {
-    // we were interrupted before, resume where we left off
-    if (m_current != null) {
-      m_stack.push(m_current);
+    isExecuting = true;
+    while (!deque.isEmpty()) {
+      final IExecutable executable = deque.pop();
+      try {
+        executable.execute(this, bridge);
+      } catch (final RuntimeException e) {
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java]: Are you sure that a RuntimeException would always be thrown if this is interrupted?

File: game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java [hunk 2]
```diff
@@ -28,37 +29,36 @@
...
   }
 
   boolean isExecuting() {
-    return m_current != null;
+    return isExecuting;
   }
 
   public boolean isEmpty() {
-    return m_stack.isEmpty();
+    return deque.isEmpty();
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java]: This could cause different behavior if its called before execute() is or after execution is interrupted and the current one is added back. Did you double check those scenarios?

File: game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java (hunk 1 above)
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java]: Well we know that it obviously can't be a checked exception.
So the only Throwables that would not be caught and rethrown here would be Errors that are usually thrown by the JVM (or AssertionErrors) and are not caught anywhere in the codebase AFAIK.
So any non-RuntimeException would instantly terminate the current thread and very likely the JVM along with it (for example in case of an OutOfMemoryError).
So I'm 99% sure we're safe here.

File: game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java (hunk 2 above)
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java]: Good point, I did not.
I'll try figuring out if there are cases when an exception is caught, where - as a response to that - executables are added to the stack, later this evening.

File: game-core/src/test/java/games/strategy/triplea/delegate/ExecutionStackTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+import games.strategy.engine.delegate.IDelegateBridge;
+
+class ExecutionStackTest {
+  private final ExecutionStack dummy = new ExecutionStack();
```
Comment (ssoloff) [inline game-core/src/test/java/games/strategy/triplea/delegate/ExecutionStackTest.java]: Sorry, forgot to add this comment to my previous review...

Curious choice for a field name.  Could you elaborate on why you chose `dummy`?  When I see something named as such, I think of a [_Dummy Object_](https://en.wikipedia.org/wiki/Test_double), but this looks like it is simply the object under test.  In that case, `executionStack` seems more appropriate.

**Comment by RoiEXLab** [comment]
@ssoloff 

> The only way I can see you could clean this up a bit (i.e. removing the `try`-`finally` that seems to be causing this discussion thread) is to use `peek()` on line 38 instead of `pop()`, and then only removing the item at the front of the deque _after_ execution succeeds.  Something like:
> 
> ```java
> deque.peek().execute(this, bridge);
> deque.pop(); // only remove if execution succeeds
> ```
> 
> However, that might cause problems if this method is re-entrant, even in a single-threaded context.

If you look at the commit history you'll see that this was the first thing I tried, because I thought the same thing too.
However it turns out that executables get added to the stack during the execution of IExecutable, so doing so will cause an infinite loop when arriving at the first battle phase in a game (i.e. nothing happens and the game doesn't proceed), because a new item is getting pushed onto the stack which is then immediately removed and the old item gets re-executed which pushes a new executable on the stack... etc.

**Comment by RoiEXLab** [comment]
@ssoloff @ron-murhammer 
Because of all the legitimate what-ifs in this PR I decided to revert back to the original code to be safe with this.
With all the added Unit Tests we get the benefit of additional Test coverage.

It'd still be nice to find out in which edge cases exactly the "redo the failed step" is really needed, so if anyone wants to investigate this, feel free to do so...

File: game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java [hunk 4]
```diff
@@ -28,37 +29,34 @@
...
-  private final Stack<IExecutable> m_stack = new Stack<>();
+  private IExecutable currentStep;
+  private final Deque<IExecutable> deque = new ArrayDeque<>();
 
   void execute(final IDelegateBridge bridge) {
     // we were interrupted before, resume where we left off
-    if (m_current != null) {
-      m_stack.push(m_current);
+    if (currentStep != null) {
+      currentStep.execute(this, bridge);
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java]: Is this change intentional?

Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/triplea/delegate/ExecutionStack.java]: It is. 
I figured it won't alter the behaviour as this item is instantly popped off the stack again and reassigned to this field again.

**Comment by ssoloff** [comment]
> If you look at the commit history you'll see that this was the first thing I tried, because I thought the same thing too.

@RoiEXLab Sorry, I should have realized that would have been the obvious first solution. :smile: Thanks for clarifying.
